library/core/src/str/mod.rs

```diff
@@ -414,12 +420,13 @@ pub fn from_utf8_mut(v: &mut [u8]) -> Result<&mut str, Utf8Error> {
 /// ```
 #[inline]
 #[stable(feature = "rust1", since = "1.0.0")]
-pub unsafe fn from_utf8_unchecked(v: &[u8]) -> &str {
-    // SAFETY: the caller must guarantee that the bytes `v`
-    // are valid UTF-8, thus the cast to `*const str` is safe.
-    // Also, the pointer dereference is safe because that pointer
-    // comes from a reference which is guaranteed to be valid for reads.
-    unsafe { &*(v as *const [u8] as *const str) }
+#[rustc_const_unstable(feature = "const_str_from_utf8_unchecked", issue = "75196")]
+#[allow(unused_attributes)]
+#[allow_internal_unstable(const_fn_union)]
+pub const unsafe fn from_utf8_unchecked(v: &[u8]) -> &str {
+    // SAFETY: the caller must guarantee that the bytes `v` are valid UTF-8.
+    // Also relies on `&str` and `&[u8]` having the same layout.
+    unsafe { StrOrSlice { slice: v }.str }
```

Comment:
I think calling `transmute` would be preferred over a hacky transmute-union.

Reply:
But, can't you use the `const_raw_ptr_deref` feature gate to keep using the old code?

Reply:
I'd prefer the transmute version

Reply:
Why? Usually a ptr cast is preferred over a transmute as it is slightly more restrictive (ensures that these are actually pointer types).

Reply:
usually I agree, but `&[u8]` and `&str` are exactly the same thing, the pointer cast and deref is harder to read than a transmute imo. But I'm fine keeping the status quo (and also using it for `as_bytes`)